### PR TITLE
Support running yggd as non-root

### PR DESCRIPTION
**Pull request by subpop**
Add support for running yggd as a non-root system user. Two new build options, "user" (defaulting to "yggdrasil") and "group" (defaulting to "yggdrasil"), are used to configure the various systemd units, systemd sysuser templates, D-Bus policy files and a pkgconfig module. This allows distributions to decide how to configure the yggdrasil service to suit their exact requirements.

Additionally, this PR includes some extra commits:

1. Separate the system and user units; prior to this, the system unit was installed twice, once to the system services directory and again into the user services directory. This proved to be inadequate. The user slice does not have a "network-online" target, resulting in the user service failing to start. Separating the two units fixes the user service so it works properly.
2. A bug was found in the way values were computed in the `constants` package when running as root compared to non-root. The commit reworks the way constants are determined; they're refactored to prefer systemd provided values, falling back to XDG paths (if non-root) and custom paths if root.
3. Use systemd macros in the spec file to properly reload services during installation.

All three of these commits are not technically part of this feature implementation, but were so closely related I felt it acceptable to keep them in this same PR.

The final commit not mentioned above implements the meson build options, systemd sysuser template and the necessary configuration changes to the systemd services and D-Bus policy files.

Fixes: #152

### CODE REVIEW

**Comment by subpop** [comment]
> #157 also modified [com.redhat.Yggdrasil1.Worker1.echo.conf](https://github.com/RedHatInsights/yggdrasil/pull/157/files#diff-7eb531046c44fae543db356a1c18ce0ad5c9217049353d04fe87eac8da7f80a4) to be able to work with yggd.service running as non-root user.

This raises an interesting question. The echo worker exists as a demonstration of how to write a worker. If we incorporate the non-root sysuser option into that worker, it means that other, out-of-tree workers would need to somehow know the sysuser running the main `yggd` process at compile time in order to write the appropriate busconfig policy. Perhaps we need to add a pkgconfig ".pc" file again so that out-of-tree workers can configure their busconfig policy files.

> Original PR https://github.com/RedHatInsights/yggdrasil/pull/157 does more things. It also introduces group ygg_worker to be able run worker as non-root user. The yggd_worker should be defined in the system user configuration file like this: https://github.com/RedHatInsights/yggdrasil/pull/157/files#diff-59036393d7e836a1742b10c66b44912b081764e45354aebee0b2c608ac03f391

This is a good point too, but is tangential to #152. Running `yggd` as a non-root user is not the same as running all workers as a non-root user. There are workers that expect to have full root access (yggdrasil-worker-package-manager, rhc-worker-playbook, rhc-script-worker), and there are others (yggdrasil-worker-desktop-notification) that expect to run as non-root. I don't think the upstream _yggdrasil_ project should dictate which user a worker runs as.

We can make `echo` run as a non-root user, again, for demonstrative purposes, but I don't think we need to add a new user for that; we can reuse the same sysuser that `yggd` uses.

**Review by jirihnidek — COMMENTED** [review]
I'm adding only notes, because this is still draft...

I would add group `ygg_worker` (or anything defined during meson build).

File: data/systemd/system/yggdrasil-bus@.service [hunk 1]
```diff
@@ -5,6 +5,7 @@ Requires=yggdrasil-bus@%i.socket
 
 [Service]
 Type=notify
+User=@user@
```
[the remark this answers is not in the document]
Comment (subpop) [reply data/systemd/system/yggdrasil-bus@.service:8]: There is no need to explicitly set the `Group=` directive. According to [the documentation](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#User=):

> If no group is set, the default group of the user is used.

Since the sysuser template creates a user and corresponding "user group" automatically, this group is automatically used by the systemd service. I confirmed this in the output of `ps`:

```
[link@lima-fedora rhc]$ ps axo pid,user,group,args | grep yggd
 472015 link     link     grep --color=auto yggd
3648112 yggdras+ yggdras+ /usr/bin/yggd
3648808 link     link     /usr/bin/yggd
```

File: data/systemd/yggdrasil-sysuser.conf.in [hunk 2]
```diff
@@ -0,0 +1 @@
+u @user@ - "yggdrasil system user" @sharedstatedir@/yggdrasil -
```
[the remark this answers is not in the document]
Comment (subpop) [reply data/systemd/yggdrasil-sysuser.conf.in:1]: Fair point about the header comment. I didn't feel it was necessary for a single line file, but it does make the file format more explicit.

According to [the documentation](https://www.freedesktop.org/software/systemd/man/latest/sysusers.d.html#Shell):

> The login shell of the user. If not specified, this will be set to /usr/sbin/nologin, except if the UID of the user is 0, in which case /bin/sh will be used.
> 
> Only applies to lines of type u and should otherwise be left unset (or "-"). It is recommended to omit this, unless a shell different /usr/sbin/nologin must be used.

I omitted it from the file specifically so that the default value is used.

File: data/dbus/yggd.conf.in [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    <allow send_destination="com.redhat.Yggdrasil1.Dispatcher1" />
+
+    <!-- Only @user@ can own the Yggdrasil1 destination name. -->
+    <allow own="com.redhat.Yggdrasil1" />
+  </policy>
+  <policy context="default">
+    <!-- Anyone can send messages to Yggdrasil destination. -->
+    <allow send_destination="com.redhat.Yggdrasil1" />
+  </policy>
+</busconfig>
```
[the remark this answers is not in the document]
Comment (subpop) [reply data/dbus/yggd.conf.in:19]: Yes, you do make some good changes to both the yggd policy and the echo policy. I need to look at those in more detail.

[the remark this answers is not in the document]
Comment (subpop) [reply data/dbus/yggd.conf.in:19]: You're definitely on to something using a group policy to restrict access to sending messages to the `com.redhat.Yggdrasil1.Dispatcher1` name. I don't know if it needs to be another group, or can the `yggdrasil` group be used for that purpose.

[the remark this answers is not in the document]
Comment (subpop) [reply data/dbus/yggd.conf.in:27]: I tweaked the bus policy a lot. I dropped the unique, per-worker one again, and added a `send_destination_prefix` attribute rule instead. I know this creates an requirement of D-Bus >= 1.14, but I think that's acceptable.

**Comment by jirihnidek** [comment]
We are limited here, because non-root user could not read consumer cert and key. We need to find some solution for this issue.

**Comment by subpop** [comment]
OK. I've experimented with this some more, and I'm of the opinion that running as a non-root user without the use of setuid/setgid syscalls is a cleaner approach. Initializing the process as root creates a window of time when the process is privileged. While it's true now that (1) we control that window; and (2) the window is very small, over time, those two statements may not remain true. Having this window of privileged execution means present creates the possibility of a privileged execution attack. Considering we can make this work as non-root from process origination, we can avoid the additional risk this theoretical attack surface brings.

File: data/dbus/yggd.conf.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (subpop) [reply data/dbus/yggd.conf.in:22]: I reworked this again. dbus-broker doesn't support `send_destination_prefix` policy attributes. ಠ_ಠ

File: meson_options.txt [hunk 4]
```diff
@@ -5,3 +5,4 @@ option('examples', type: 'boolean', value: false, description: 'Build and instal
 option('gobuildflags', type: 'array', value: ['-buildmode', 'pie'], description: 'Additional build flags to be passed to the Go compiler')
 option('goldflags', type: 'string', value: '', description: 'Additional linker flags to be passed to the Go compiler')
 option('build_srpm', type: 'boolean', value: false, description: 'Enable SRPM builds')
+option('user', type: 'string', value: 'yggdrasil', description: 'User to run system yggd as')
```
[the remark this answers is not in the document]
Comment (subpop) [reply meson_options.txt]: I added an explicit group option too. This allows distributions to configure a separate username and group pair, should that suit their configuration better.

File: data/systemd/yggdrasil-sysuser.conf.in [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#Type Name     ID   GECOS                   Home directory             Shell
+u     @user@   -    "yggdrasil system user" @sharedstatedir@/yggdrasil -
+g     @group@  -
+m     @user@   @group@
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply data/systemd/yggdrasil-sysuser.conf.in:4]: The output of `packit build local` says: `ignoring unsupported sysuser type: m` despite installed version of systemd should support it according documentation: https://www.freedesktop.org/software/systemd/man/latest/sysusers.d.html

Or is it bug in some rpm macro? Like this: https://github.com/rpm-software-management/rpm/blob/master/macros.in#L1341 Weird.

[the remark this answers is not in the document]
Comment (subpop) [reply data/systemd/yggdrasil-sysuser.conf.in:4]: The builds in COPR that packit triggered don't seem to have that same warning line. I couldn't find it in [RHEL 8](https://download.copr.fedorainfracloud.org/results/packit/RedHatInsights-yggdrasil-224/rhel-8-x86_64/07321759-yggdrasil/builder-live.log.gz)

[the remark this answers is not in the document]
Comment (subpop) [reply data/systemd/yggdrasil-sysuser.conf.in:4]: I just tried building locally with `packit build in-mock`, and the RPMs built fine.

[the remark this answers is not in the document]
Comment (jirihnidek) [reply data/systemd/yggdrasil-sysuser.conf.in:4]: I was able to build the RPM after restarting of system. When I installed the RPM to the system, and I tried to start the service, then start of `yggdrasil.service` was terminated with following error:

```
root@thinkpad-p1:/home/jhnidek/github/redhat-insights/yggdrasil# systemctl status yggdrasil.service 
○ yggdrasil.service - yggdrasil system service
     Loaded: loaded (/usr/lib/systemd/system/yggdrasil.service; disabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: inactive (dead)
       Docs: https://github.com/RedHatInsights/yggdrasil
root@thinkpad-p1:/home/jhnidek/github/redhat-insights/yggdrasil# systemctl start yggdrasil.service 
Job for yggdrasil.service failed because the control process exited with error code.
See "systemctl status yggdrasil.service" and "journalctl -xeu yggdrasil.service" for details.
root@thinkpad-p1:/home/jhnidek/github/redhat-insights/yggdrasil# systemctl status yggdrasil.service 
× yggdrasil.service - yggdrasil system service
     Loaded: loaded (/usr/lib/systemd/system/yggdrasil.service; disabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: failed (Result: exit-code) since Mon 2024-05-06 14:07:11 CEST; 8s ago
       Docs: https://github.com/RedHatInsights/yggdrasil
    Process: 80880 ExecStart=/usr/bin/yggd (code=exited, status=1/FAILURE)
   Main PID: 80880 (code=exited, status=1/FAILURE)
        CPU: 153ms

kvě 06 14:07:11 thinkpad-p1 systemd[1]: Starting yggdrasil.service - yggdrasil system service...
kvě 06 14:07:11 thinkpad-p1 yggd[80880]: [yggd] 2024/05/06 14:07:11 /home/jhnidek/github/redhat-insights/yggdrasil/builddir/dist/srpm/yggdrasil-0.4.1/_build/src/github.com/redhatinsights/yggdrasil/cmd/yggd/m>
kvě 06 14:07:11 thinkpad-p1 yggd[80880]: [yggd] 2024/05/06 14:07:11 /home/jhnidek/github/redhat-insights/yggdrasil/builddir/dist/srpm/yggdrasil-0.4.1/_build/src/github.com/redhatinsights/yggdrasil/internal/w>
kvě 06 14:07:11 thinkpad-p1 yggd[80880]: cannot setup client: cannot connect client: cannot connect dispatcher: cannot request name on bus: Request to own name refused by policy
kvě 06 14:07:11 thinkpad-p1 systemd[1]: yggdrasil.service: Main process exited, code=exited, status=1/FAILURE
kvě 06 14:07:11 thinkpad-p1 systemd[1]: yggdrasil.service: Failed with result 'exit-code'.
kvě 06 14:07:11 thinkpad-p1 systemd[1]: Failed to start yggdrasil.service - yggdrasil system service.
```

It seems that some policy file is not enough... I will add comment to the code.

File: data/dbus/yggd.conf.in [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE busconfig PUBLIC "-//freedesktop//DTD D-BUS Bus Configuration 1.0//EN" "https://dbus.freedesktop.org/doc/busconfig.dtd">
+<busconfig>
+  <policy user="@user@">
```
Comment (jirihnidek) [inline data/dbus/yggd.conf.in:4]: Following line is missing:
```
<allow own="com.redhat.Yggdrasil1"/>
```
Otherwise, it is not possible to start the service.

File: data/dbus/yggd.conf.in [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    <!-- Only @user@ can own the Dispatcher1 destination name. -->
+    <allow own="com.redhat.Yggdrasil1.Dispatcher1" />
+
+    <!-- Only @user@ can own the Yggdrasil1 destination name. -->
+    <allow own="com.redhat.Yggdrasil1" />
```
Comment (jirihnidek) [inline data/dbus/yggd.conf.in:9]: It seems that this line does not work here.

File: data/dbus/yggd.conf.in (hunk 6 above)
Comment (subpop) [reply data/dbus/yggd.conf.in:4]: Oh interesting! I had not considered whether the order in which lines appear matters.

Comment (subpop) [reply data/dbus/yggd.conf.in:4]: I switched the order of these two `<allow>` tags.

File: worker/echo/systemd/com.redhat.Yggdrasil1.Worker1.echo.service.in [hunk 8]
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=yggdrasil echo worker service
+Documentation=https://github.com/RedHatInsights/yggdrasil
+
+[Service]
+Type=dbus
+User=@user@
+Group=@group@
```
Comment (jirihnidek) [inline worker/echo/systemd/com.redhat.Yggdrasil1.Worker1.echo.service.in:8]: ... I can see that you use same service user and group for running echo worker service. Shouldn't we have special user & group for echo worker? Something similar I have my in my origin PR: https://github.com/RedHatInsights/yggdrasil/pull/157/files#diff-7eb531046c44fae543db356a1c18ce0ad5c9217049353d04fe87eac8da7f80a4

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Overall good. I have only few suggestions and questions.

File: dist/srpm/yggdrasil.spec.in [hunk 9]
```diff
@@ -66,6 +66,14 @@ BuildRequires:  golang >= 1.18
 
 %description %{common_description}
 
+%package devel
+Summary:        %{name} development files
+
+%description devel
+%{common_description}
+
+Contains files needed for yggdrasil worker development.
```
[the remark this answers is not in the document]
Comment (subpop) [reply dist/srpm/yggdrasil.spec.in:75]: The package you're looking at is the one that's autogenerated by the go-srpm
macros. If this was an upstream Go package with source code, those source files
should end up in that package.

The package you commented on is different. If you look in the list of built
RPMs, there are two packages with the "-devel" suffix:

```
Wrote: /builddir/build/SRPMS/yggdrasil-0.4.1-0.20240508git7276b50.fc39.src.rpm
Wrote: /builddir/build/RPMS/yggdrasil-debugsource-0.4.1-0.20240508git7276b50.fc39.aarch64.rpm
Wrote: /builddir/build/RPMS/golang-github-redhatinsights-yggdrasil-devel-0.4.1-0.20240508git7276b50.fc39.noarch.rpm
Wrote: /builddir/build/RPMS/yggdrasil-devel-0.4.1-0.20240508git7276b50.fc39.aarch64.rpm
Wrote: /builddir/build/RPMS/yggdrasil-0.4.1-0.20240508git7276b50.fc39.aarch64.rpm
Wrote: /builddir/build/RPMS/yggdrasil-debuginfo-0.4.1-0.20240508git7276b50.fc39.aarch64.rpm
```

The package description you commented on is the package `yggdrasil-devel`. This
package contains only one file (for now), but is only relevant to yggdrasil
developers, so it should be included in a "-devel" subpackage.

```
$ rpm -qpil /var/lib/mock/fedora-39-aarch64/result/yggdrasil-devel-0.4.1-0.20240508git7276b50.fc39.aarch64.rpm
Name        : yggdrasil-devel
Version     : 0.4.1
Release     : 0.20240508git7276b50.fc39
Architecture: aarch64
Install Date: (not installed)
Group       : Unspecified
Size        : 189
License     : GPL-3.0-only
Signature   : (none)
Source RPM  : yggdrasil-0.4.1-0.20240508git7276b50.fc39.src.rpm
Build Date  : Wed 08 May 2024 10:24:38 AM EDT
Build Host  : lima-fedora.local
URL         : https://github.com/redhatinsights/yggdrasil
Summary     : yggdrasil development files
Description :

yggdrasil is a system daemon that subscribes to topics on an MQTT broker and
routes any data received on the topics to an appropriate child "worker" process,
exchanging data with its worker processes through a D-Bus message broker.

Contains files needed for yggdrasil worker development.
/usr/lib64/pkgconfig/yggdrasil.pc
```

File: data/systemd/yggdrasil-sysuser.conf.in (hunk 5 above)
[the remark this answers is not in the document]
Comment (subpop) [reply data/systemd/yggdrasil-sysuser.conf.in:4]: The purpose of the group is to permit processes to send to the
`com.redhat.Yggdrasil1.Dispatcher1` destination. What is the advantage of making
workers use a separate group from the main process?

```xml
  <policy group="@group@">
    <!-- Only members of the @group@ group can send messages to Dispatcher1
    destination. -->
    <allow send_destination="com.redhat.Yggdrasil1.Dispatcher1" />
  </policy>
```

File: worker/echo/systemd/com.redhat.Yggdrasil1.Worker1.echo.service.in (hunk 8 above)
Comment (subpop) [reply worker/echo/systemd/com.redhat.Yggdrasil1.Worker1.echo.service.in:8]: What advantage does it bring to run workers as separate users & groups?

**Comment by subpop** [comment]
Some of my individual comments seem to have been lost between Visual Studio Code and GitHub. They all revolve around the same subject though. You suggested using a separate group for worker processes. What is the reason for having two groups: `yggdrasil` and `yggdrasil_workers`? The reason a group exists at all is to permit processes to send to the `com.redhat.Yggdrasil1.Dispatcher1` destination. I don't see why that group needs to be separate from the primary group of the main `yggd` process.

Both `user` and `group` are configurable build options, so if a distribution wanted to use a separate group, they could configure it like so: `meson setup -Duser=yggdrasil -Dgroup=yggdrasil_workers`

The [sysusers.d manpage](https://man7.org/linux/man-pages/man5/sysusers.d.5.html) implies that in this case, a separate primary group and secondary group will get created. If the user and group value match, a single user & group entry will be created. If they differ, then a primary group matching the user name will be created, and a secondary group will be created and the user will be added to the group.

When configured with a matching user and group the following user & group configuration is created:

> Creating group 'yggdrasil' with GID 991.
> Creating user 'yggdrasil' (yggdrasil system user) with UID 991 and GID 991.

```
$ id yggdrasil
uid=991(yggdrasil) gid=991(yggdrasil) groups=991(yggdrasil)
```

When configured with a separate user & group, the following user & group configuration is created:

> Creating group 'yggdrasil_workers' with GID 991.
> Creating group 'yggdrasil' with GID 987.
> Creating user 'yggdrasil' (yggdrasil system user) with UID 987 and GID 987.

```
$ id yggdrasil
uid=987(yggdrasil) gid=987(yggdrasil) groups=987(yggdrasil),991(yggdrasil_workers)
```

So should a distribution wish to run workers with a separate group, they could configure it to do so. I don't see the need to add an explicit, second "worker group" in this case, since the sysusers configuration file we ship can already handle both cases.

File: dist/srpm/yggdrasil.spec.in (hunk 9 above)
[the remark this answers is not in the document]
Comment (jirihnidek) [reply dist/srpm/yggdrasil.spec.in:75]: Oh, you are right. I overlooked this devel package. It generates too many devel packages. :-)

File: worker/echo/systemd/com.redhat.Yggdrasil1.Worker1.echo.service.in (hunk 8 above)
Comment (jirihnidek) [reply worker/echo/systemd/com.redhat.Yggdrasil1.Worker1.echo.service.in:8]: It is simple. When worker runs as a same user as yggd, then it has the same access permissions. Example: do we want to allow to any worker to own `com.redhat.Yggdrasil1` D-Bus interface and block starting of yggdrasil service? I guess we do not want.

Of course, it is possible to run this example worker as the same user/group as we use for yggdrasil service and simplify few bits, but I think it is not good to show bad practice in any example. Another example: The MySQL server does not use the same user/group as Apache server use, despite database server "communicates" with HTTP server.

**Comment by jirihnidek** [comment]
@subpop if the only one intention of `yggdrasil` group is to allow workers to call `Transmit()` method from interface `com.redhat.Yggdrasil1.Dispatcher1`, then it is little bit risky. Why? Same group is used for creating all files. There is danger that we will give write access to some file that should not be writable by workers. Maybe it is not so risky now, but you never know what file will be created by yggdrasil service in future.

For this reason I still believe that it is good to have separate group allowing to worker to call method(s) from interface `com.redhat.Yggdrasil1.Dispatcher1`

**Comment by subpop** [comment]
> @subpop if the only one intention of `yggdrasil` group is to allow workers to call `Transmit()` method from interface `com.redhat.Yggdrasil1.Dispatcher1`, then it is little bit risky. Why? The same group is also used for creating all files. There is danger that we will give write access to some file that should not be writable by workers. Maybe it is not so risky now, but you never know what file will be created by yggdrasil service in the future.
> 
> For this reason I still believe that it is good to have separate group allowing to worker to call method(s) from interface `com.redhat.Yggdrasil1.Dispatcher1`

That's a very good point. I'll add a second group for workers to run as.

File: data/systemd/yggdrasil-sysuser.conf.in [hunk 10]
```diff
@@ -0,0 +1,4 @@
+#Type Name            ID   GECOS                   Home directory             Shell
+u     @user@          -    "yggdrasil system user" @sharedstatedir@/yggdrasil -
+u     @worker_user@   -    "yggdrasil worker user" -                          -
+m     @user@          @worker_user@
```
Comment (jirihnidek) [inline data/systemd/yggdrasil-sysuser.conf.in:4]: It is correct, but it looks strange. This adds user to supplementary group `@worker_user@` (replaced by default by `yggdrasil-worker`).

I think that we should move echo worker to its own RPM package, and it will solve some issues. We can do this in another PR.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
We are almost there. I have problem with starting echo worker using `systemctl`. I suspect that there has to be some mistake in some configuration file.

Output of commands:

```
root@thinkpad-p1:/home/jhnidek/github/redhat-insights/yggdrasil# systemctl start com.redhat.Yggdrasil1.Worker1.echo.service
Job for com.redhat.Yggdrasil1.Worker1.echo.service failed because the control process exited with error code.
See "systemctl status com.redhat.Yggdrasil1.Worker1.echo.service" and "journalctl -xeu com.redhat.Yggdrasil1.Worker1.echo.service" for details.

root@thinkpad-p1:/home/jhnidek/github/redhat-insights/yggdrasil# systemctl status com.redhat.Yggdrasil1.Worker1.echo.service
× com.redhat.Yggdrasil1.Worker1.echo.service - yggdrasil echo worker service
     Loaded: loaded (/usr/lib/systemd/system/com.redhat.Yggdrasil1.Worker1.echo.service; disabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: failed (Result: exit-code) since Wed 2024-05-22 14:15:42 CEST; 10s ago
       Docs: https://github.com/RedHatInsights/yggdrasil
    Process: 133660 ExecStart=/usr/libexec/yggdrasil/echo (code=exited, status=1/FAILURE)
   Main PID: 133660 (code=exited, status=1/FAILURE)
        CPU: 7ms

kvě 22 14:15:42 thinkpad-p1 systemd[1]: Starting com.redhat.Yggdrasil1.Worker1.echo.service - yggdrasil echo worker service...
kvě 22 14:15:42 thinkpad-p1 echo[133660]: 2024/05/22 14:15:42 error: cannot connect: cannot request name on bus: Request to own name refused by policy
kvě 22 14:15:42 thinkpad-p1 systemd[1]: com.redhat.Yggdrasil1.Worker1.echo.service: Main process exited, code=exited, status=1/FAILURE
kvě 22 14:15:42 thinkpad-p1 systemd[1]: com.redhat.Yggdrasil1.Worker1.echo.service: Failed with result 'exit-code'.
kvě 22 14:15:42 thinkpad-p1 systemd[1]: Failed to start com.redhat.Yggdrasil1.Worker1.echo.service - yggdrasil echo worker service.
```

File: worker/echo/dbus/com.redhat.Yggdrasil1.Worker1.echo.service.in
Comment (jirihnidek) [inline worker/echo/dbus/com.redhat.Yggdrasil1.Worker1.echo.service.in:1]: There are two service files for the same echo worker service and I believe that this file should be dropped, because another service file https://github.com/RedHatInsights/yggdrasil/pull/224/files#diff-94dead0b1d43ad053b75aa7355b91ed63ed9fa3623f3fc9e26781fce0ccae97f is correct. It runs the echo worker using service user and it is compatible with installed policy file.

When this file exists, then it is not possible to start the service due to policy file. I tested that deleting this file fixes the issue described in this comment: https://github.com/RedHatInsights/yggdrasil/pull/224#pullrequestreview-2071037944

File: worker/echo/dbus/meson.build [hunk 11]
```diff
@@ -0,0 +1,15 @@
+configure_file(
+  configuration: config_data,
+  input: 'com.redhat.Yggdrasil1.Worker1.echo.service.in',
+  output: '@BASENAME@',
+  install: true,
+  install_dir: dbus.get_variable(pkgconfig: 'system_bus_services_dir')
+)
```
Comment (jirihnidek) [inline worker/echo/dbus/meson.build:7]: If the service file from this directory was deleted, then the configuration file should also not be installed.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
I think that I have solution for the last issue. Look at my comments...

**Review by jirihnidek — APPROVED** [review]
OK, to move forward I'm approving this PR. We can solve remaining bits in another PR(s)
